net/netns/netns_linux.go

```diff
@@ -26,19 +26,49 @@ import (
 // wgengine/router/router_linux.go.
 const tailscaleBypassMark = 0x80000
 
-// ipRuleOnce is the sync.Once & cached value for ipRuleAvailable.
-var ipRuleOnce struct {
+// socketMarkWorksOnce is the sync.Once & cached value for useSocketMark.
+var socketMarkWorksOnce struct {
 	sync.Once
 	v bool
 }
 
-// ipRuleAvailable reports whether the 'ip rule' command works.
+// socketMarkWorks returns whether SO_MARK works. If unsure, it returns true as the
+// consequences of returning an incorrect false are worse.
+func socketMarkWorks() bool {
+	addr, err := net.ResolveUDPAddr("udp", "127.0.0.1:1")
+	if err != nil {
+		return true
+	}
+
+	sConn, err := net.DialUDP("udp", nil, addr)
+	if err != nil {
+		return true
+	}
+	defer sConn.Close()
+
+	rConn, err := sConn.SyscallConn()
+	if err != nil {
+		return true
+	}
+
+	var sockErr error
+	err = rConn.Control(func(fd uintptr) {
+		sockErr = setBypassMark(fd)
+	})
+	if err != nil || sockErr != nil {
+		return false
+	}
+
+	return true
+}
+
+// useSocketMark reports whether SO_MARK works.
 // If it doesn't, we have to use SO_BINDTODEVICE on our sockets instead.
-func ipRuleAvailable() bool {
```

Comment:
Why replace `ipRuleAvailable` ? If it's not available, regardless of whether we can set socket marks, marked packets won't route correctly.

Reply:
Because it is not sufficient: Azure and AWS App Runner run containers which do not have CAP_NET_ADMIN. They *do* have `ip rule` and can print the table of existing rules, but attempting to use SO_MARK fails with EPERM. So on those platforms we decide to use SO_MARK, which fails, and then Tailscale cannot make any data connections.

If we try to use SO_MARK as proposed here, then we detect that SO_MARK doesn't work and we use SO_BINDTODEVICE instead. SO_BINDTODEVICE does work, even without CAP_NET_ADMIN.

Reply:
Sorry, I meant: SO_MARK by itself is not sufficient, so the check needs to be additive. If `ip rule` isn't there, we can't add the plumbing that makes the marks work in the first place. Checking both SGTM